### PR TITLE
Refactor TasksService and tests

### PR DESCRIPTION
**Pull request by Dorvaryn**
#### Problem
The tasksService is getting big and complex to read through.

#### Solution
We can refactor it a bit, removing some non needed abstractions (good by `IfThenMap`) and clean up the order of methods. This should make it a bit more manageable. 
We can also use Kotlin's ability to declare functions with back ticks to write more explicit tests in order to guide peoples discovery of the code base.

##### Screenshots
| Before | After |
| ------ | ----- |
|before|after|

##### Test(s) added 
no since no behaviour was changed, tests are renamed though.

###### Paired with 
No-one

### CODE REVIEW

File: Architecture/TodoApp/android/core/src/main/java/com/novoda/todoapp/tasks/service/PersistedTasksService.java [hunk 1]
```diff
@@ -388,9 +350,9 @@ public void delete(final Task task) {
         final long deleteActionTimestamp = clock.timeInMillis();
         remoteDataSource.deleteTask(task.id())
                 .compose(deleteLocallyThenConfirmOrMarkAsError(task, deleteActionTimestamp))
-                .flatMap(updateAndPersistIfDeleteActionIsMostRecent(task, deleteActionTimestamp))
+                .filter(actionIsAbsentOrNoMoreRecentThan(task.id(), deleteActionTimestamp))
```
Comment (danybony) [inline Architecture/TodoApp/android/core/src/main/java/com/novoda/todoapp/tasks/service/PersistedTasksService.java]: 👍

File: Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/statistics/presenter/StatisticsPresenterTest.kt [hunk 2]
```diff
@@ -35,7 +35,7 @@ class StatisticsPresenterTest {
     }
 
     @Test
-    fun given_ThePresenterIsPresenting_on_EmissionOfNewIdleEventWithStatistics_it_ShouldPresentTheStatisticsToTheView() {
+    fun `Given the presenter is presenting, On emission of a new idle event with statistics, It should present the statistics to the displayer`() {
```
Comment (danybony) [inline Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/statistics/presenter/StatisticsPresenterTest.kt]: Nice! ✨

File: Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/statistics/presenter/StatisticsPresenterTest.kt [hunk 3]
```diff
@@ -134,7 +134,7 @@ class StatisticsPresenterTest {
     }
 
     @Test
-    fun given_ThePresenterIsNotPresenting_on_EmissionOfNewErrorEventWithNoStatistics_it_ShouldNotPresentAnythingToTheView() {
+    fun `Given the presenter is not presenting, On emission of a new error loading event, It should not present anything to the displayer`() {
```
Comment (danybony) [inline Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/statistics/presenter/StatisticsPresenterTest.kt]: On emission of a new **empty** error loading event

File: Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt [hunk 4]
```diff
@@ -437,7 +419,7 @@ class TasksPresenterTest {
     }
 
     @Test
-    fun given_ThePresenterIsPresenting_on_StopPresenting_it_ShouldDetachTheNavDrawerActionListener_FromTheDisplayer() {
+    fun `Given the presenter is presenting, On stopPresenting, It should detach the action listener from the topLevelMenuDisplayer`() {
         givenThePresenterIsNotPresenting()
```
Comment (danybony) [inline Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt]: `givenThePresenterIsPresenting()`

File: Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt [hunk 5]
```diff
@@ -83,7 +83,7 @@ class TasksPresenterTest {
     }
 
     @Test
-    fun given_ThePresenterIsPresenting_on_EmissionOfALoadingEventWithNoData_it_ShouldPresentTheLoadingScreen() {
+    fun `Given the presenter is presenting, On emission of an empty loading event, It should present to the loading screen`() {
```
Comment (danybony) [inline Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt]: It should present ~~to~~ the loading screen

File: Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt [hunk 6]
```diff
@@ -92,7 +92,7 @@ class TasksPresenterTest {
     }
 
     @Test
-    fun given_ThePresenterIsPresenting_on_EmissionOfALoadingEventWithData_it_ShouldPresentTheLoadingIndicator() {
+    fun `Given the presenter is presenting, On emission of a loading event with data, It should present to the loading indicator`() {
```
Comment (danybony) [inline Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt]: It should present ~~to~~ the loading indicator

File: Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt [hunk 7]
```diff
@@ -102,7 +102,7 @@ class TasksPresenterTest {
 
 
     @Test
-    fun given_ThePresenterIsPresenting_on_EmissionOfLoadingEventWithSomeDeletedData_it_ShouldPresentLoadingIndicator() {
+    fun `Given the presenter is presenting, On emission of a loading event with some deleted data, It should present to the loading indicator`() {
```
Comment (danybony) [inline Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt]: It should present ~to~ the loading indicator

File: Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt [hunk 8]
```diff
@@ -111,7 +111,7 @@ class TasksPresenterTest {
     }
 
     @Test
-    fun given_ThePresenterIsPresenting_on_EmissionOfLoadingEventWithAllDeletedData_it_ShouldPresentLoadingScreen() {
+    fun `Given the presenter is presenting, On emission of a loading event with all deleted data, It should present to the loading screen`() {
```
Comment (danybony) [inline Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt]: It should present ~~to~~ the loading screen

File: Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt [hunk 9]
```diff
@@ -120,7 +120,7 @@ class TasksPresenterTest {
     }
 
     @Test
-    fun given_TheCurrentFilterIsAll_on_EmissionOfAnIdleEventWithNoData_it_ShouldPresentTheEmptyScreen() {
+    fun `Given the current filter is all, On emission of an empty idle event, It should present to the empty screen`() {
```
Comment (danybony) [inline Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/presenter/TasksPresenterTest.kt]: It should present ~~to~~ the empty screen

File: Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/service/PersistedTasksServiceTest.kt [hunk 10]
```diff
@@ -126,7 +126,7 @@ class PersistedTasksServiceTest {
     }
 
     @Test
-    fun given_TheLocalDataIsEmptyAndRemoteFails_on_GetTasksEvents_it_ShouldReturnError() {
+    fun `Given the local data is empty and remote fails, On getTasksEvent, It should return and error`() {
```
Comment (danybony) [inline Architecture/TodoApp/android/core/src/test/kotlin/com/novoda/todoapp/tasks/service/PersistedTasksServiceTest.kt]: It should return **an** error

**Comment by danybony** [comment]
The first part of the PR is difficult to follow, but good refactor 👍 
Also great test names! (:heart: kotlin)
Just some comments and LGTM 👍

**Comment by frapontillo** [comment]
Gooooood stuff! 👍

**Pull request merged**